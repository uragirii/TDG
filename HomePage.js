import React, { Component } from 'react'
import {  View, FlatList, SafeAreaView} from 'react-native'
import { Caption, Headline, Subheading, ActivityIndicator } from "react-native-paper";
import { Post } from "./components/Post";

export class HomePage extends Component {
    state = {
        loading : true,
        allPosts : [],
        postPerPage : 10,     // Customize prefrence
        error: "",
        loadingMore:false
    }
    componentDidMount(){
        // Call the API here. As this API does not support lazy loading of 
        // elements in group of 10. Load 100, and then recall the API
        // as user scrolls through 100
        fetch("https://jsonplaceholder.typicode.com/posts")
            .then(res => res.json())
            .then(res =>{
                // Contains API response
                // For complex Data handling, Redux can also be used 
                this.setState({
                    allPosts: res.slice(0,this.state.postPerPage),
                    error : "",
                    loading:false
                })
            })
            .catch(err =>{
                this.setState({
                    loading:false,
                    error: err
                })
            })
    }
    getMoredata(){
        this.setState({loadingMore:true})
        fetch("https://jsonplaceholder.typicode.com/posts")
            .then(res => res.json())
            .then(res =>{
                // Contains API response
                // For complex Data handling, Redux can also be used 
                // if the data gets too large, crop upper data
                this.setState({
                    allPosts: this.state.allPosts.concat(res.slice(0,this.state.postPerPage)),
                    error : "",
                    loadingMore:false
                })
            })
            .catch(err=>{
                console.log(err)
                this.setState({
                    loadingMore:false,
                    error:err
                })
            })
    }
    render() {
        if (this.state.loading){
            return (
                    <View style={{flexDirection: "row", justifyContent: "center",height:"100%", backgroundColor:"#081631"}}>
                        <ActivityIndicator size="small" color="#f90024" />
                        <Subheading style={{alignSelf:"center", padding:"5%", color:"white"}}>Loading Posts</Subheading>
                    </View>
            )
        }
        else{
            return (
                <SafeAreaView style={{height:"100%", backgroundColor:"#081631"}} >
                    <FlatList
                        data = {this.state.allPosts}
                        initialNumToRender={10}
                        keyExtractor={(item, index)=>index.toString()}
                        onEndReachedThreshold={0.6}
                        onEndReached={()=>{
                            this.getMoredata()
                        }}
                        ListHeaderComponent={()=>{
                            return (
                            <Headline 
                                style={{
                                    color:"white", 
                                    fontSize:45,
                                    paddingTop:"15%",
                                    margin:"5%",
                                    paddingBottom:"7%"
                                }} 
                                numberOfLines={1} 
                                adjustsFontSizeToFit>Posts</Headline>)
                        }}
                        ListFooterComponent={() =>{
                            if(this.state.error===""){
                                return (
                                    <View style={{flexDirection: "row", justifyContent: "center", alignItems: "center",marginBottom:"2%", paddingVertical:"5%"}}>
                                        <ActivityIndicator size="small" color="#f90024"/>
                                        <Caption style={{color:"white", padding:"5%"}}>Loading More Posts</Caption>
                                    </View>
                                )
                            }
                            else{
                                return (
                                    <View style={{flexDirection: "row", justifyContent: "center", alignItems: "center",marginBottom:"2%", paddingVertical:"5%"}}>
                                        <Caption style={{color:"white", padding:"5%"}}>Error Occured while loading</Caption>
                                    </View>
                                )
                            }
                            
                        }}
                        ListFooterComponentStyle={{  padding:"5%"}}
                        renderItem={({item, index})=>{
                            return (
                                <Post
                                    title ={item.title}
                                    body = {item.body}
                                />
                            )
                        }}
                    />
                </SafeAreaView>
            )
        }
    }
}

export default HomePage
