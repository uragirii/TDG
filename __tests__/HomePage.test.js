/**
 * @format
 */

import 'react-native';
import React from 'react';
import {HomePage} from '../HomePage';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

fetch = jest.fn(() => Promise.resolve());

it('renders correctly', () => {
  const tree = renderer.create(<HomePage/>).toJSON()
  expect(tree).toMatchSnapshot();
});
