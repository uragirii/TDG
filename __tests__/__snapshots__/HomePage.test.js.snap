// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#081631",
      "flexDirection": "row",
      "height": "100%",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "height": 24,
          "opacity": 1,
          "width": 24,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          style={
            Object {
              "height": 24,
              "transform": Array [
                Object {
                  "rotate": "45deg",
                },
              ],
              "width": 24,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "height": 12,
                "overflow": "hidden",
                "width": 24,
              }
            }
          >
            <View
              style={
                Object {
                  "height": 24,
                  "transform": Array [
                    Object {
                      "translateY": 0,
                    },
                    Object {
                      "rotate": "-165deg",
                    },
                  ],
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 12,
                    "overflow": "hidden",
                    "width": 24,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "borderColor": "#f90024",
                      "borderRadius": 12,
                      "borderWidth": 2.4,
                      "height": 24,
                      "width": 24,
                    }
                  }
                />
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          style={
            Object {
              "height": 24,
              "transform": Array [
                Object {
                  "rotate": "45deg",
                },
              ],
              "width": 24,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "height": 12,
                "overflow": "hidden",
                "top": 12,
                "width": 24,
              }
            }
          >
            <View
              style={
                Object {
                  "height": 24,
                  "transform": Array [
                    Object {
                      "translateY": -12,
                    },
                    Object {
                      "rotate": "345deg",
                    },
                  ],
                  "width": 24,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 12,
                    "overflow": "hidden",
                    "width": 24,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "borderColor": "#f90024",
                      "borderRadius": 12,
                      "borderWidth": 2.4,
                      "height": 24,
                      "width": 24,
                    }
                  }
                />
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
          "textAlign": "left",
        },
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontFamily": "System",
            "fontWeight": "400",
            "textAlign": "left",
            "writingDirection": "ltr",
          },
          Array [
            Object {
              "fontSize": 16,
              "letterSpacing": 0.5,
              "lineHeight": 24,
              "marginVertical": 2,
            },
            Object {
              "alignSelf": "center",
              "color": "white",
              "padding": "5%",
            },
          ],
          Array [
            Object {
              "fontSize": 16,
              "letterSpacing": 0.5,
              "lineHeight": 24,
              "marginVertical": 2,
            },
            Object {
              "alignSelf": "center",
              "color": "white",
              "padding": "5%",
            },
          ],
        ],
      ]
    }
  >
    Loading Posts
  </Text>
</View>
`;
